File: Backtracking/9663번 - N-Queen/9663.cpp
#include <iostream>
#include <vector>

using namespace std;

int n;
int col[16] = { 0, };
int ans = 0;

void input() {
    cin >> n;
}

// 해당 칸의 비숍을 놓을 때 같은 열에 비숍이 있는지 또는 대각선에 비숍이 있는지 확인
bool check(int x) {
    for(int i = 0; i < x; i++) {
        if(col[x] == col[i] || abs(col[x] - col[i]) == x - i) {
            return false;
        }
    }
    
    return true;
}

void back(int x) {
    if(x == n) {
        ans++;
        return;
    }
    
    // 각 행에서 비숍을 놓음.
    for(int i = 0; i < n; i++) {
        col[x] = i;
        
        // 해당 칸에 비숍을 놓을 수 있는지 확인
        if(check(x)) {
            back(x + 1);
        }
    }
}

void solve() {
    back(0);
    
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Backtracking/9663번 - N-Queen/9663.py
import sys

input = sys.stdin.readline

n = int(input())
ans = 0

def back(x, col):
    global ans

    if x == n:
        ans += 1
        return

    for i in range(n):
        col[x] = i
        chk = True
        
        for j in range(x):
            if col[x] == col[j] or abs(col[x] - col[j]) == x - j:
                chk = False
                break
        
        if chk:
            back(x + 1, col)

def solve():
    back(0, [0] * n)
    
    print(ans)
    

if __name__ == '__main__':
    solve()